AppController/djinn.rb

```diff
@@ -4155,148 +4143,97 @@ def rsync_files(dest_node)
-    # TODO: this should be the class or module
-    table = @options['table']
-    # require db_file
-    begin
-      require "#{table}_helper"
-    rescue => e
-      backtrace = e.backtrace.join("\n")
-      HelperFunctions.log_and_crash("Unable to find #{table} helper." +
-        " Please verify datastore type: #{e}\n#{backtrace}")
-    end
-
-    @nodes.each { |node|
-      master_ip = node.private_ip if node.jobs.include?("db_master")
-      unless slave_ips.include? node.private_ip
-        slave_ips << node.private_ip if node.jobs.include?("db_slave")
-      end
-    }
-
-    Djinn.log_debug("Master is at #{master_ip}, slaves are at #{slave_ips.join(', ')}")
+    taskqueue_ips = []
 
     my_public = my_node.public_ip
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/my_public_ip", "#{my_public}\n")
-
     my_private = my_node.private_ip
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/my_private_ip", "#{my_private}\n")
-
-    head_node_ip = get_shadow.public_ip
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/head_node_ip", "#{head_node_ip}\n")
-
-    login_ip = @options['login']
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/login_ip", "#{login_ip}\n")
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/masters", "#{master_ip}\n")
 
-    if @nodes.length  == 1
-      Djinn.log_info("Only saw one machine, therefore my node is " +
-        "also a slave node")
-      slave_ips = [ my_private ]
-    end
-
-    slave_ips_newlined = slave_ips.join("\n")
-    HelperFunctions.write_file("#{APPSCALE_CONFIG_DIR}/slaves", "#{slave_ips_newlined}\n")
+    # Populate the appropriate list.
+    @nodes.each { |node|
+      all_ips << node.private_ip
+      load_balancer_ips << node.private_ip if node.is_load_balancer?
+      master_ips << node.private_ip if node.is_db_master?
+      memcache_ips << node.private_ip if node.is_memcache?
+      search_ips = node.private_ip if node.is_search?
+      slave_ips << node.private_ip if node.is_db_slave?
+      taskqueue_ips << node.private_ip if node.is_taskqueue_master? ||
+        node.is_taskqueue_slave?
+    }
+    slave_ips << master_ips[0] if slave_ips.empty?
+
+    # Turn the arrays into string.
+    all_ips_content = all_ips.join("\n") + "\n"
+    memcache_content = memcache_ips.join("\n") + "\n"
+    load_balancer_content = load_balancer_ips.join("\n") + "\n"
+    taskqueue_content = taskqueue_ips.join("\n") + "\n"
+    login_content = login_ips.join("\n") + "\n"
+    master_content = master_ips.join("\n") + "\n"
+    search_content = slave_ips.join("\n") + "\n"
+    slaves_content = slave_ips.join("\n") + "\n"
+
+    new_content = all_ips_content + login_content + load_balancer_content +
+      master_content + memcache_content + my_public + my_private +
+      num_of_nodes + taskqueue_content + search_content + slaves_content
+
+    # If nothing changed since last time we wrote locations file(s), skip it.
+    if new_content != @locations_content
```

Comment:
This is still going to be true because the locations_content will have the shuffled taskqueue_ips... If you sorted both taskqueue lists and added it as an "and" to the condition that would test for that. I'm not really concerned about it too much its just not saving us anything right now.